extensions/amp-form/0.1/amp-form.js

```diff
@@ -177,8 +185,69 @@ export class AmpForm {
 
 
 /**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
+  const inputs = toArray(form.querySelectorAll('input,select,textarea'));
+  for (let i = 0; i < inputs.length; i++) {
+    if (!inputs[i].checkValidity()) {
+      return reportInputValidity(inputs[i]);
+    }
+  }
+}
+
+
+/**
+ * Revalidates the currently focused input after a change.
+ * @param {!KeyboardEvent} event
+ * @private
+ */
+function onInvalidInputKeyUp_(event) {
+  validationBubble.hide();
+  if (!event.target.checkValidity()) {
+    validationBubble.show(event.target, event.target.validationMessage);
+  }
+}
+
+
+/**
+ * Hides validation bubble and removes listeners on the invalid input.
+ * @param {!Event} event
+ * @private
+ */
+function onInvalidInputBlur_(event) {
+  validationBubble.hide();
+  event.target.removeEventListener('blur', onInvalidInputBlur_);
+  event.target.removeEventListener('keyup', onInvalidInputKeyUp_);
+}
+
+
+/**
+ * Focuses and reports the invalid message of the input in a message bubble.
+ * @param {!HTMLInputElement} input
+ * @private
+ */
+function reportInputValidity(input) {
+  input./*REVIEW*/focus();
```

Comment:
Done. I've wrapped the original call of reportFormValidation with the vsync.mutate call. Is that alright? Otherwise I'd have to pass down the window or the vsync service instance to these functions.